addons/website_sale/models/product.py

```diff
@@ -170,6 +178,44 @@ def _compute_parents_and_self(self):
             else:
                 category.parents_and_self = category
 
+    def _get_categories_and_count_from_product_domain(self, categ_domain):
+        """ Return a tuple with the main product.public.category and a dict with category id as key and product count as value """
+        Product = self.env['product.template']
+        where_query = Product._where_calc(categ_domain)
+        Product._apply_ir_rules(where_query)
+        from_clause, where_clause, args = where_query.get_sql()
+        query = """
+            SELECT c.id, COUNT(DISTINCT product_template)
+            FROM %s
+            INNER JOIN %s ON product_template_id = product_template.id
+            INNER JOIN %s c ON product_public_category_id = c.id
+            WHERE %s
+            GROUP BY c.id
+        """ % (self._fields['product_tmpl_ids'].relation, Product._table, self._table, where_clause)
+        self.flush()
+        self.env.cr.execute(query, args)
+        categs_nbr_prod = dict(self.env.cr.fetchall())
+        categs_ids = list(categs_nbr_prod.keys())
+        categs = self.search([('id', 'parent_of', categs_ids), ('parent_id', '=', False)])
+
+        def add_nb_child(node, categs_nbr_prod):
+            """ Recursively add the number of product in childs to there parents.
+                Done this way for performance. Selecting the childs of the categories in SQL is an issue """
+            if not categs_nbr_prod.get(node.id):
+                # Only child with products in them are returned by sql.
+                categs_nbr_prod[node.id] = 0
+            for child in node.child_id:
```

Comment:
Since you are already reading `.child_id` here, I don't think the `with_prefetch` will work in the controller, it is already too late. It should be done before calling this.

Reply:
If all categories that are iterated here are not in `categs_ids` it will just not work to set that as the prefetch. What has to be done is find all the parent/children in one query before, and then filter on parent_if = False, just like the code was previously.